### ISSUE TITLE
--no-bail behavior inconsistent

### ISSUE THREAD

**Issue by c-vetter** [pnpm/pnpm#8013]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

9.0.6

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/c-vetter/repro_pnpm_inconsistent-nobail

### Reproduction steps

In the repro, `pnpm lint`, then check the exit code.

### Describe the Bug

`--no-bail` is listed on the [website for `--recursive`](https://pnpm.io/cli/recursive#--no-bail), but not for [plain `run`](https://pnpm.io/cli/run).

Based on #3036, `--no-bail` is currently specced to always exit 0, at least without `--recursive`. I don't have a monorepo to test right now, but if I read the tests correctly, with `--recursive`, `--no-bail` exits with a fixed code, with the symbol `"ERR_PNPM_RECURSIVE_FAIL"`. That would mean that the behavior for `pnpm run --recursive --no-bail` is consistent with the website docs, while the behavior for `pnpm run --no-bail` is different.

#7381 is related. I have pretty much exactly the same use-case as @mizdra and am coming from `npm-run-all` as well. Based on the current behavior, I cannot switch although I would like to drop the dependency. I was thinking about adding this there, but it seemed more appropriate to open a new bug report.

#7381 could be interpreted as advocating a new parameter, in addition to `--no-bail`. While I don't think that is actually the case, I want to be clear. `--no-bail` should be fixed, there is no need for another parameter – that would just add confusion without benefit. For those users who don't care about the exit code, it's trivial to ignore it. But when the failure state matters, we cannot easily extract that without parsing the logs, which is just the wrong way to go.

### Expected Behavior

When running with `--no-bail`, exit code should be 0 if all sub-runs exit 0, and non-0 otherwise. Irrespective of whether or not `--recursive` is active.

### Which Node.js version are you using?

20.12.2

### Which operating systems have you used?

- [ ] macOS
- [X] Windows
- [ ] Linux

**Comment by cruzdanilo** [comment]
the same issue happens for `pnpm test`

**Comment by DanilAgafonov** [comment]
I've been investigating that for a bit and it all comes from this: https://github.com/pnpm/pnpm/blob/bffe53896bd2292699d44601efd37c2f9326b7ba/exec/plugin-commands-script-runners/src/run.ts#L280

It works when `stdio = "pipe"`, but doesn't with `stdio = "inherit"` (when concurrency = 1).

In my case I use one CPU core on CI and that's causing that issue. The problem here, I can not force it to use `pipe` mode. Even if I pass `--workspace-concurrency=4`, [this part](https://github.com/pnpm/pnpm/blob/bffe53896bd2292699d44601efd37c2f9326b7ba/config/config/src/concurrency.ts#L29) would not allow it on CI.

**Comment by askoufis** [comment]
I'm seeing a similar issue. Using `pnpm --no-bail run '/^lint:.*/'` to run multiple `lint:` scripts in parallel, I get an exit code of 0 even when one of the lint scripts fail.

**Comment by antitoxic** [comment]
We are facing the same problem that @askoufis describes. 

This has no workaround. Even worse, even if I try to force and run as `--recursive` when `--filter` resolves to a single package with scripts matching the regex, it follows the non-recursive "run multiple matched scripts in one package" path so in this case **`--no-bail` will act as non-recursive even if I pass `--recursive` explicitly**. 

The only place the main inconsistency is documented for `--no-bail` in non-recursive runs is here in the source code: https://github.com/pnpm/pnpm/blob/089484aca81f6624b2a7f636476aaa9219ded33f/exec/commands/src/run.ts#L140-L141

@zkochan is it possible to make `--no-bail` act the same in both recursive and non-recursive? Even if that is hidden behind some new config property to avoid breaking changes? Alternatively at least some workaround? I tried my best even with lllm models help but no real solution. I'm guessing any change that would be done is around exit code tracking here: https://github.com/pnpm/pnpm/blob/089484aca81f6624b2a7f636476aaa9219ded33f/exec/commands/src/run.ts#L295-L312 to make it more like https://github.com/pnpm/pnpm/blob/089484aca81f6624b2a7f636476aaa9219ded33f/exec/commands/src/runRecursive.ts#L161-L172

Right now, because of this we are rethinking using `pnpm run /regex/` but we are heavily invested into it.

**Comment by antitoxic** [comment]
> Below is some AI-assisted proposal, i'm not opening a PR since I just want to know if that's a direction that maintainers are ok going to:

## Proposed Fix for pnpm Source Code

### Summary

The fix adds a new `--report-failure` CLI flag (aliased as `--continue-on-error` for discoverability) to `pnpm run`. When `--no-bail` is combined with `--report-failure`, pnpm will:
1. Run all matched scripts to completion (existing `--no-bail` behavior)
2. Exit with non-zero if **any** script failed (new behavior, opt-in)

This mirrors what `runRecursive` already does with `throwOnCommandFail`, but for the non-recursive regex script matching path.

---

### Files to Change

#### 1. `exec/plugin-commands-script-runners/src/run.ts`

**a) Add `report-failure` to CLI options:**

```typescript
// In cliOptionsTypes():
export function cliOptionsTypes (): Record<string, unknown> {
  return {
    ...pick([
      'bail',
      'sort',
      'unsafe-perm',
      'use-node-version',
      'workspace-concurrency',
      'scripts-prepend-node-path',
    ], allTypes),
    ...IF_PRESENT_OPTION,
    recursive: Boolean,
    reverse: Boolean,
    'resume-from': String,
    'report-summary': Boolean,
    'reporter-hide-prefix': Boolean,
    'report-failure': Boolean,       // <-- ADD THIS
  }
}
```

**b) Add help text:**

```typescript
// In help(), add to the options list:
{
  description: 'Used with --no-bail. When set, the process exits with a non-zero exit code if any of the scripts fail. Without this flag, --no-bail always exits with code 0.',
  name: '--report-failure',
},
```

**c) Add to `RunOpts` type:**

```typescript
export type RunOpts =
  & Omit<Config, 'dir'>
  & { recursive?: boolean }
  & Pick<Config, 'dir'>
  & (
    | { recursive?: false } & Partial<Pick<Config, 'allProjects' | 'selectedProjectsGraph' | 'workspaceDir'>>
    | { recursive: true } & Required<Pick<Config, 'allProjects' | 'selectedProjectsGraph' | 'workspaceDir'>>
  )
  & {
    argv?: {
      original: string[]
    }
    fallbackCommandUsed?: boolean
    reportFailure?: boolean          // <-- ADD THIS
  }
  & CheckDepsStatusOptions
```

**d) Replace the broken `try/catch` block (the core fix):**

Replace this (around line 270-282):

```typescript
  try {
    const limitRun = pLimit(concurrency)

    const runScriptOptions: RunScriptOptions = {
      enablePrePostScripts: opts.enablePrePostScripts ?? false,
      syncInjectedDepsAfterScripts: opts.syncInjectedDepsAfterScripts,
      workspaceDir: opts.workspaceDir,
    }
    const _runScript = runScript.bind(null, { manifest, lifecycleOpts, runScriptOptions, passedThruArgs })

    await Promise.all(specifiedScripts.map(script => limitRun(() => _runScript(script))))
  } catch (err: unknown) {
    if (opts.bail !== false) {
      throw err
    }
  }
  return undefined
```

With this:

```typescript
  const limitRun = pLimit(concurrency)

  const runScriptOptions: RunScriptOptions = {
    enablePrePostScripts: opts.enablePrePostScripts ?? false,
    syncInjectedDepsAfterScripts: opts.syncInjectedDepsAfterScripts,
    workspaceDir: opts.workspaceDir,
  }
  const _runScript = runScript.bind(null, { manifest, lifecycleOpts, runScriptOptions, passedThruArgs })

  if (opts.bail !== false) {
    // Default behavior: fail fast on first error
    await Promise.all(specifiedScripts.map(script => limitRun(() => _runScript(script))))
  } else {
    // --no-bail: run all scripts to completion regardless of failures
    const results = await Promise.allSettled(
      specifiedScripts.map(script => limitRun(() => _runScript(script)))
    )

    if (opts.reportFailure) {
      const failures = results.filter(
        (r): r is PromiseRejectedResult => r.status === 'rejected'
      )
      if (failures.length > 0) {
        throw new PnpmError(
          'RUN_NO_BAIL_FAILURES',
          `Some scripts failed: ${failures.length} of ${specifiedScripts.length}`,
          {
            hint: failures
              .map((f) => f.reason?.message ?? String(f.reason))
              .join('\n'),
          }
        )
      }
    }
  }
  return undefined
```

**Key changes explained:**
- `Promise.all` → `Promise.allSettled` when `--no-bail` is active. This ensures **all** scripts run to completion regardless of failures (fixing the subtle issue where `Promise.all` short-circuits on first rejection and leaves other scripts in a detached state).
- After all scripts settle, if `--report-failure` is set, collect failures and throw a `PnpmError` with code `ERR_PNPM_RUN_NO_BAIL_FAILURES` (pnpm auto-prepends `ERR_PNPM_`).

---

#### 2. `config/config/src/Config.ts`

Add to the `Config` interface:

```typescript
export interface Config extends OptionsFromRootManifest {
  // ... existing fields ...
  reportFailure?: boolean   // <-- ADD THIS (used with --no-bail to report failures)
}
```

---

#### 3. `config/config/src/types.ts`

Add to `pnpmTypes`:

```typescript
export const pnpmTypes = {
  // ... existing fields ...
  'report-failure': Boolean,  // <-- ADD THIS
}
```

---

#### 4. Test file: `exec/plugin-commands-script-runners/test/run.ts`

Add a test:

```typescript
test('pnpm run with regex, --no-bail and --report-failure exits non-zero when a script fails', async () => {
  prepare({
    scripts: {
      'lint:a': 'exit 0',
      'lint:b': 'exit 1',
      'lint:c': 'exit 0',
    },
  })

  let error: PnpmError | undefined
  try {
    await run.handler({
      dir: process.cwd(),
      bail: false,
      reportFailure: true,
      // ... other required opts
    }, ['/^lint:/'])
  } catch (err: any) {
    error = err
  }

  expect(error).toBeDefined()
  expect(error?.code).toBe('ERR_PNPM_RUN_NO_BAIL_FAILURES')
})

test('pnpm run with regex, --no-bail and --report-failure exits zero when all scripts pass', async () => {
  prepare({
    scripts: {
      'lint:a': 'exit 0',
      'lint:b': 'exit 0',
    },
  })

  // Should not throw
  await run.handler({
    dir: process.cwd(),
    bail: false,
    reportFailure: true,
    // ... other required opts
  }, ['/^lint:/'])
})

test('pnpm run with regex and --no-bail without --report-failure exits zero (backward compat)', async () => {
  prepare({
    scripts: {
      'lint:a': 'exit 0',
      'lint:b': 'exit 1',
    },
  })

  // Should NOT throw — preserves existing behavior
  await run.handler({
    dir: process.cwd(),
    bail: false,
    reportFailure: false,
    // ... other required opts
  }, ['/^lint:/'])
})
```

**Comment by karlhorky** [comment]
@antitoxic seems like a PR would be a better, more reviewable way of making that proposal

**Comment by antitoxic** [comment]
@karlhorky sure, that was my next step but wanted to gather feedback here so I don't load yet another PR (there are 200+ PRs). 

I just don't understand if there  was intent in making the current bail behaviour inconsistent in the first place. Like maybe I'm missing some background.

Also not sure about the suggested AI argument names, especially the concept of aliasing to something  that seems to duplicate --no-bail. Etc.

**Comment by antitoxic** [comment]
@karlhorky @zkochan: https://github.com/pnpm/pnpm/pull/12263
